### PR TITLE
Add valid_move? method to pawn

### PR DESCRIPTION
**Pull request by jonmast**
Thanks to @radiofreemattd for helping clean this up.

### CODE REVIEW

File: app/models/pawn.rb [hunk 1]
```diff
@@ -1,2 +1,64 @@
...
+  def y_move_factor
+    moved? ? POST_INITIAL_MOVE_FACTOR : INITIAL_MOVE_FACTOR
+  end
+
+  def moved?
+    updated_at != created_at
+  end
+
+  def forward_move?(destination_y)
+    y_distance = destination_y - position_y
```
Comment (radiofreemattd) [inline app/models/pawn.rb]: We could call `diff_in_y` if we want to be consistent here, I think.

File: app/models/pawn.rb [hunk 2]
```diff
@@ -1,2 +1,64 @@
...
+    if color == 'white'
+      y_distance > 0
+    elsif color == 'black'
+      y_distance < 0
+    end
```
Comment (radiofreemattd) [inline app/models/pawn.rb]: It's starting to feel like `chess_piece.rb` should have a method to return the opposite color.

File: app/models/pawn.rb [hunk 3]
```diff
@@ -1,2 +1,64 @@
 class Pawn < ChessPiece
+  INITIAL_MOVE_FACTOR = 2
+  POST_INITIAL_MOVE_FACTOR = 1
+
+  def valid_move?(destination_x, destination_y)
+    return false unless forward_move?(destination_y)
+    if vertical_move?(destination_x, destination_y)
+      valid_vertical_move?(destination_x, destination_y)
+    else
```
Comment (radiofreemattd) [inline app/models/pawn.rb]: Do we need this? I think we want both method calls within here, don't we?

File: app/models/pawn.rb (hunk 2 above)
Comment (radiofreemattd) [inline app/models/pawn.rb]: Again, we can probably use `diff_in_y` here for consistency.

File: app/models/pawn.rb (hunk 1 above)
Comment (jonmast) [inline app/models/pawn.rb]: diff_in_y always returns a positive number.

Comment (radiofreemattd) [inline app/models/pawn.rb]: You're right :+1:

File: app/models/pawn.rb [hunk 4]
```diff
@@ -1,2 +1,64 @@
 class Pawn < ChessPiece
+  INITIAL_MOVE_FACTOR = 2
```
Comment (mscoutermarsh) [inline app/models/pawn.rb]: How about
`FIRST_MOVE_FACTOR`
`SECOND_MOVE_FACTOR`

File: app/models/pawn.rb [hunk 5]
```diff
@@ -1,2 +1,64 @@
...
+  POST_INITIAL_MOVE_FACTOR = 1
+
+  def valid_move?(destination_x, destination_y)
+    return false unless forward_move?(destination_y)
+    if vertical_move?(destination_x, destination_y)
+      valid_vertical_move?(destination_x, destination_y)
+    else
+      valid_capture?(destination_x, destination_y)
+    end
```
Comment (mscoutermarsh) [inline app/models/pawn.rb]: looks really good. could remove the else and end with `valid_capture?(destination_x, destination_y`). And add a return before `valid_vertical_move`

File: app/models/pawn.rb [hunk 6]
```diff
@@ -1,2 +1,64 @@
...
+
+  private
+
+  def valid_vertical_move?(destination_x, destination_y)
+    return false if moved_to_far?(destination_y)
+    return false if position_occupied?(destination_x, destination_y)
+    !obstructed?(destination_x, destination_y)
+  end
+
+  def moved_to_far?(destination_y)
```
Comment (mscoutermarsh) [inline app/models/pawn.rb]: `moved_too_far?`

File: app/models/pawn.rb [hunk 7]
```diff
@@ -1,2 +1,64 @@
...
+    moved? ? POST_INITIAL_MOVE_FACTOR : INITIAL_MOVE_FACTOR
+  end
+
+  def moved?
+    updated_at != created_at
+  end
+
+  def forward_move?(destination_y)
+    y_distance = destination_y - position_y
+    if color == 'white'
```
Comment (mscoutermarsh) [inline app/models/pawn.rb]: i'm not positive. But I think since color is a enum. you can use `white?`

File: spec/models/pawn_spec.rb [hunk 8]
```diff
@@ -4,4 +4,58 @@
...
+    context 'when color is white' do
+      let(:pawn) { create(:pawn, position_x: 4, position_y: 4, color: :white) }
+
+      it 'returns true for one step forward moves' do
+        expect(pawn.valid_move?(4, 5)).to eq true
+      end
+
+      it 'returns false for two step forward moves if not first move' do
+        pawn.move_to!(4, 5)
+        expect(pawn.valid_move?(4, 7)).to eq false
```
Comment (mscoutermarsh) [inline spec/models/pawn_spec.rb]: nice!

File: spec/models/pawn_spec.rb [hunk 9]
```diff
@@ -4,4 +4,58 @@
...
+      it 'returns false for three step forward moves' do
+        expect(pawn.valid_move?(4, 7)).to eq false
+      end
+
+      it 'returns false for diagonal moves' do
+        expect(pawn.valid_move?(5, 5)).to eq false
+      end
+
+      it 'returns true for diagonal captures' do
+        create(:pawn, position_x: 5, position_y: 5, color: :black, game: pawn.game)
```
Comment (mscoutermarsh) [inline spec/models/pawn_spec.rb]: Does having game set here matter?

Comment (mscoutermarsh) [inline spec/models/pawn_spec.rb]: nvmind, I got it! makes sense

**Comment by mscoutermarsh** [comment]
looks great! merge it whenever you are ready :)

**Pull request merged**